### PR TITLE
[Auto Parallel] Upgrade fluid comm operators to be compatible with new comm library

### PR DESCRIPTION
**Pull request by GhostScreaming**
### PR types
Others

### PR changes
Others

### Description
Pcard-70448

Based on compatible new communication library upgrade PR https://github.com/PaddlePaddle/Paddle/pull/56604. Old communication library is used in default, export `FLAGS_dynamic_static_unified_comm=1` to use new communication library.

Following operators are upgraded compatibly  now: `alltoall`, `barrier`, `c_concat`, `c_allreduce`, `c_sync_stream_op`, `c_wait_comm_op`, `c_wait_compute_op`, `recv_v2_op`, `send_v2_op`, `c_reduce_op`, `c_allgather_op`, `c_scatter_op`, `c_reducescatter_op`.

### CODE REVIEW

File: python/paddle/distributed/auto_parallel/static/process_group.py [hunk 1]
```diff
@@ -194,28 +213,29 @@ def instantiate(self):
...
-                    alltoall_tmp, 'use_calc_stream', True, 'ring_id', ring_id
-                )
-                paddle.device.cuda.synchronize()
-
-        if self.nranks > 1:
-            barrier_tensor = paddle.full([1], 1, dtype="int32")
-            paddle._legacy_C_ops.barrier(
-                barrier_tensor, barrier_tensor, 'ring_id', 0
-            )
+     #      barrier_tensor = paddle.full([1], 1, dtype="int32")
```
Comment (hitywt) [inline python/paddle/distributed/auto_parallel/static/process_group.py]: 这里注释可以取消掉测试下，注释掉可能会跳过一些问题

File: paddle/phi/core/distributed/store/tcp_store.cc [hunk 2]
```diff
@@ -351,12 +351,20 @@ TCPStore::TCPStore(std::string host,
...
+  VLOG(0) << "input timeout" << timeout << ", member timeout:" << _timeout;
   if (_is_master) {
+    VLOG(0) << "before master port " << port << ", num_workers:" << num_workers
+            << ", timeout: " << timeout;
     _server = detail::TCPServer::create(port, num_workers, timeout);
+    VLOG(0) << "after master port " << port << ", num_workers:" << num_workers
+            << ", timeout: " << timeout;
   }
 
+  VLOG(0) << "bfore connect host " << host << ", port:" << port;
```
Comment (hitywt) [inline paddle/phi/core/distributed/store/tcp_store.cc]: VLOG(0)都删除下吧

File: paddle/fluid/operators/collective/barrier_op.cu.cc [hunk 3]
```diff
@@ -38,13 +40,45 @@ class BarrierOpCUDAKernel : public framework::OpKernel<T> {
...
-    auto stream = ctx.cuda_device_context().stream();
-    ncclRedOp_t nccl_red_type = ncclSum;
-    PADDLE_ENFORCE_GPU_SUCCESS(platform::dynload::ncclAllReduce(
-        sendbuff, recvbuff, numel, dtype, nccl_red_type, comm->comm(), stream));
-    platform::GpuStreamSync(stream);
+    const auto& comm_context_manager =
+        phi::distributed::CommContextManager::GetInstance();
+    VLOG(0) << "new comm_context_manager has rid " << rid;
+    if (comm_context_manager.Has(std::to_string(rid))) {
+      VLOG(0) << "new comm_context_manager has rid " << rid;
```
Comment (hitywt) [inline paddle/fluid/operators/collective/barrier_op.cu.cc]: 这里VLOG(0)也是调试用的，可以删除

File: paddle/phi/core/distributed/comm_context_manager.cc [hunk 4]
```diff
@@ -64,7 +73,61 @@ void CommContextManager::CreateNCCLCommContext(
...
+    dev_ctx->SetHostAllocator(phi::memory_utils::GetHostAllocator());
+    dev_ctx->SetZeroAllocator(
+        phi::memory_utils::GetZeroAllocator(CommContextManager::device_id));
+    dev_ctx->SetHostZeroAllocator(phi::memory_utils::GetHostZeroAllocator());
+    dev_ctx->SetPinnedAllocator(phi::memory_utils::GetPinnedAllocator());
+    dev_ctx->PartialInitWithAllocator();
+    // *******************************************************************************
+    // dev_ctx->SetAllocator(
+    //     paddle::memory::allocation::AllocatorFacade::Instance()
+    //         .GetAllocator(phi::GPUPlace(CommContextManager::device_id),
```
Comment (hitywt) [inline paddle/phi/core/distributed/comm_context_manager.cc]: 注释代码可以删除了

File: paddle/phi/core/distributed/nccl_comm_context.cc [hunk 5]
```diff
@@ -37,6 +37,36 @@ NCCLCommContext::NCCLCommContext(int rank, int size, ncclUniqueId nccl_id)
 
 ncclComm_t NCCLCommContext::GetNcclComm() { return nccl_comm_; }
 
+gpuStream_t NCCLCommContext::GetStream() {
+  if (!dev_ctx_) {
+    VLOG(0) << "NCCLCommContext::GetStream dev_ctx nullptr";
```
Comment (hitywt) [inline paddle/phi/core/distributed/nccl_comm_context.cc]: 调试VLOG(0)需要删除

File: python/paddle/distributed/auto_parallel/static/process_group.py [hunk 6]
```diff
@@ -58,6 +59,7 @@ def remove_process_group(ring_id):
 def new_process_group(
     ranks, group_id=None, force_new_group=False, group_type=None
 ):
+    logger.info(f"debug new_process_group group_id: {group_id}")
```
Comment (hitywt) [inline python/paddle/distributed/auto_parallel/static/process_group.py]: 调试log，可以删除

File: test/collective/test_collective_sendrecv.py [hunk 7]
```diff
@@ -25,18 +25,36 @@ class TestSendRecvOp(TestDistBase):
...
+    # def test_sendrecv_dynamic_shape(self):
+    #     self.check_with_place(
+    #         "collective_sendrecv_op_dynamic_shape.py", "sendrecv_dynamic_shape"
+    #     )
+
+    # def test_sendrecv_array(self):
+    #     self.check_with_place(
+    #         "collective_sendrecv_op_array.py", "sendrecv_array"
+    #     )
+
```
Comment (hitywt) [inline test/collective/test_collective_sendrecv.py]: 这里为什么删除了

**Review by hitywt — APPROVED** [review]
LGTM

File: test/collective/test_collective_sendrecv.py (hunk 7 above)
Comment (GhostScreaming) [reply test/collective/test_collective_sendrecv.py]: 之前的注释忘记删了，已经修复

**Review by zhiqiu — DISMISSED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for flags

**Review by XieYunshen — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**